File: backend/graphBuilder.js
function haversine([lon1, lat1], [lon2, lat2]) {
  const toRad = (d) => d * Math.PI / 180;
  const R = 6371; 
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a = Math.sin(dLat/2)**2 + Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon/2)**2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}
// module.exports = graphBuilder;
function graphBuilder(coords, trafficData = {}) {
  const graph = {};

  for (let i = 0; i < coords.length - 1; i++) {
      const from = coords[i];
      const to = coords[i + 1];

      const baseDistance = haversine(from, to);
      const edgeKey = `${from.join(",")}_${to.join(",")}`;

      const trafficMultiplier = trafficData?.[edgeKey] ?? 1;

      const weight = baseDistance * trafficMultiplier;

      const fromKey = from.join(",");
      const toKey = to.join(",");

      if (!graph[fromKey]) graph[fromKey] = [];
      if (!graph[toKey]) graph[toKey] = [];

      graph[fromKey].push({ node: toKey, weight });
      graph[toKey].push({ node: fromKey, weight }); // IMPORTANT
  }

  return graph;
}
module.exports = graphBuilder;

File: backend_and_ML/utils/graph_builder.py
import math
from ML.traffic_predictor import get_traffic_multiplier

def haversine(p1, p2):
    lon1, lat1 = p1
    lon2, lat2 = p2

    R = 6371
    to_rad = lambda d: d * math.pi / 180

    dlat = to_rad(lat2 - lat1)
    dlon = to_rad(lon2 - lon1)

    a = (
        math.sin(dlat / 2) ** 2 +
        math.cos(to_rad(lat1)) *
        math.cos(to_rad(lat2)) *
        math.sin(dlon / 2) ** 2
    )

    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

def graph_builder(coords):
    graph = {}
    traffic_multiplier = get_traffic_multiplier()
    for i in range(len(coords) - 1):
        frm, to = coords[i], coords[i + 1]
        base_distance = haversine(frm, to)
        weight = base_distance * traffic_multiplier
        fk = ",".join(map(str, frm))
        tk = ",".join(map(str, to))
        graph.setdefault(fk, []).append({
            "node": tk,
            "weight": weight
        })
        graph.setdefault(tk, []).append({
            "node": fk,
            "weight": weight
        })
    return graph
